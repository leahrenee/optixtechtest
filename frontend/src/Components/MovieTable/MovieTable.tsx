import { useEffect, useState } from "react";
import {
  Alert,
  Box,
  CircularProgress,
  IconButton,
  Modal,
  Snackbar,
  Table,
  TableBody,
  TableCell,
  TableContainer,
  TableHead,
  TableRow,
  TableSortLabel,
  Typography,
  styled,
  useMediaQuery,
  useTheme,
} from "@mui/material";
import Paper from "@mui/material/Paper";
import { useGetMovies } from "../../hooks/useGetMovies/useGetMovies";
import { useGetMovieCompanies } from "../../hooks/useGetMovieCompanies/useGetMovieCompanies";
import { Movie, MovieCompany } from "../../types";
import RefreshIcon from "@mui/icons-material/Refresh";
import ReviewForm from "../ReviewForm/ReviewForm";
import ErrorScreen from "../ErrorScreen/ErrorScreen";

const classes = {
  tableContainer: "movie-table-container",
  tableHeader: "movie-table-header",
};

type SortType = "desc" | "asc" | undefined;

const modalStyle = {
  position: "absolute",
  top: "50%",
  left: "50%",
  transform: "translate(-50%, -50%)",
  width: "80vw",
  bgcolor: "white",
  borderRadius: "5px",
  boxShadow: 24,
  padding: "20px 0px",
};

const StyledPaper = styled(Paper)(({ theme }) => ({
  [`& .${classes.tableContainer}`]: {
    display: "flex",
    flexDirection: "column",
    alignItems: "center",
  },

  [`& .${classes.tableHeader}`]: {
    display: "flex",
    flexDirection: "row",
    alignItems: "center",
    justifyContent: "space-between",
    padding: "2px 12px",
    borderBottom: `1px solid ${theme.palette.divider}`,
  },
}));

const MovieTable = () => {
  const [selectedMovie, setSelectedMovie] = useState<Movie | undefined>();
  const [errorShown, setErrorShown] = useState(false);
  const [loadingShown, setloadingShown] = useState(false);
  const [tableSort, setTableSort] = useState<SortType>(undefined);
  const [modalOpen, setModalOpen] = useState(false);
  const [formResponseAlert, setFormResponseAlert] = useState("");

  const theme = useTheme();
  const desktop = useMediaQuery(theme.breakpoints.up("sm"));
  const mobile = useMediaQuery(theme.breakpoints.down("sm"));

  // Data hooks
  const {
    movies,
    isLoading: isLoadingMovies,
    error: moviesError,
    refetchMovies,
  } = useGetMovies();

  const {
    movieCompanies,
    isLoading: isLoadingMovieCompanies,
    error: movieCompaniesError,
    refetchMovieCompanies,
  } = useGetMovieCompanies();

  useEffect(() => {
    if (movieCompaniesError || moviesError) {
      setErrorShown(true);
    }
  }, [movieCompaniesError, moviesError]);

  useEffect(() => {
    if (isLoadingMovies || isLoadingMovieCompanies) {
      setloadingShown(true);
    } else {
      setloadingShown(false);
    }
  }, [isLoadingMovies, isLoadingMovieCompanies]);

  useEffect(() => {
    if (selectedMovie && mobile) {
      setModalOpen(true);
    } else {
      setModalOpen(false);
    }
  }, [mobile, desktop, selectedMovie]);

  const handleRefreshClick = async () => {
    setSelectedMovie(undefined);
    setErrorShown(false);
    refetchMovies();
    refetchMovieCompanies();
  };

  const getMovieScore = (movie: Movie) => {
    return movie.reviews.reduce(
      (acc: any, i: any) => (acc + i) / movie.reviews.length,
      0
    );
  };

  const handleSort = () => {
    if (tableSort === "asc" || tableSort === undefined) {
      setTableSort("desc");
      movies.sort((a, b) => getMovieScore(a) - getMovieScore(b));
    } else if (tableSort === "desc") {
      setTableSort("asc");
      movies.sort((a, b) => getMovieScore(b) - getMovieScore(a));
    }
  };

  return (
    <>
      {loadingShown && !errorShown ? (
        <Box padding="40px" data-testid="loading-spinner">
          <CircularProgress size={60} />
        </Box>
      ) : (
        <StyledPaper>
          {errorShown ? (
            <ErrorScreen
              movieCompaniesError={movieCompaniesError}
              moviesError={moviesError}
              handleRefreshClick={handleRefreshClick}
              errorShown={errorShown}
            />
          ) : (
            <>
              <Box className={classes.tableHeader}>
                {movies.length > 1 ? (
                  <Typography>Found {movies.length} Movies</Typography>
                ) : (
                  <Typography>No Movies!</Typography>
                )}
                <IconButton
                  onClick={handleRefreshClick}
                  disabled={isLoadingMovies || isLoadingMovieCompanies}
                  color="primary"
                  data-testid="refresh-button"
                >
                  <RefreshIcon />
                </IconButton>
              </Box>
              <TableContainer className={classes.tableContainer}>
                <Table>
                  <TableHead>
                    <TableRow>
                      <TableCell>Title</TableCell>
                      <TableCell>
                        <TableSortLabel
                          direction={tableSort}
                          onClick={handleSort}
                          data-testid="sort-movie-table"
                        >
                          Review Score
                        </TableSortLabel>
                      </TableCell>
                      <TableCell>Film Company</TableCell>
                    </TableRow>
                  </TableHead>
                  <TableBody>
                    {movies.map((movie: Movie, index: number) => {
                      return (
                        <TableRow
                          hover
                          key={movie.id}
                          onClick={() => {
                            setSelectedMovie(movie);
                            mobile && setModalOpen(true);
                          }}
                          selected={selectedMovie?.id === movie.id}
                          sx={{ cursor: "pointer" }}
                          data-testid={`movie-${index}-row`}
                        >
                          <TableCell>{movie.title}</TableCell>
                          <TableCell data-testid={`movie-${index}-score`}>
                            {getMovieScore(movie).toFixed(1)}
                          </TableCell>
                          <TableCell>
                            {
                              movieCompanies.find(
                                (company: MovieCompany) =>
                                  company.id === movie.filmCompanyId
                              )?.name
                            }
                          </TableCell>
                        </TableRow>
                      );
                    })}
                  </TableBody>
                </Table>
              </TableContainer>
              {selectedMovie && (
                <>
                  {desktop ? (
                    <ReviewForm
                      selectedMovie={selectedMovie}
                      setSelectedMovie={setSelectedMovie}
                      handleFormSubmission={setFormResponseAlert}
                    />
                  ) : (
                    <Modal
                      open={modalOpen}
                      onClose={() => {
                        setModalOpen(!modalOpen);
                        setSelectedMovie(undefined);
                      }}
                    >
                      <Box sx={modalStyle}>
                        <ReviewForm
                          selectedMovie={selectedMovie}
                          setSelectedMovie={setSelectedMovie}
                          handleFormSubmission={setFormResponseAlert}
                        />
                      </Box>
                    </Modal>
                  )}
                </>
              )}
            </>
          )}
          <Snackbar
            anchorOrigin={{ horizontal: "center", vertical: "bottom" }}
            open={!!formResponseAlert}
            autoHideDuration={2000}
          >
            <Alert severity="success" variant="filled">
              {formResponseAlert}
            </Alert>
          </Snackbar>
        </StyledPaper>
      )}
    </>
  );
};

export default MovieTable;
