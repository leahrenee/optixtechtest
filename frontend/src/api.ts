const baseUrl = "http://localhost:3000";

export const getData = async <T>(path: string): Promise<T> => {
  let data;

  try {
    const res = await fetch(`${baseUrl}/${path}`);

    if (res.ok) {
      data = await res.json();
    } else {
      if (res.status === 404) throw new Error("404, Not found");
      if (res.status === 500) throw new Error("500, internal server error");
      throw new Error(`Server responded with ${res.status}`);
    }
  } catch (error) {
    throw new Error(`Fetching failed: ${error}`);
  }

  return data;
};

export const postData = async <T, R>(path: string, data: T): Promise<R> => {
  let reviewResponse;

  try {
    const res = await fetch(`${baseUrl}/${path}`, {
      method: "POST",
      headers: {
        "content-type": "application/json",
        body: JSON.stringify(data),
      },
    });

    if (res.ok) {
      reviewResponse = await res.json();
    } else {
      if (res.status === 404) throw new Error("404, Not found");
      if (res.status === 500) throw new Error("500, internal server error");
      throw new Error(`Server responded with ${res.status}`);
    }
  } catch (error) {
    console.error("Fetch", error);
  }

  return reviewResponse;
};
