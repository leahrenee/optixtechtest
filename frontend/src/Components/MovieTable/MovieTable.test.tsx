import MovieTable from "./MovieTable";
import {
  act,
  fireEvent,
  getByTestId,
  render,
  waitFor,
} from "@testing-library/react";
import { getData } from "../../api";
import { mockMovieCompanyData, mockMoviesData } from "../../testData";

jest.mock("../../api");
const mockGetData = getData as jest.Mock;

// Adding this promise to wait for the hardcoded 2 second loading delay
const wait = () => new Promise((resolve) => setTimeout(resolve, 2200));

describe("MovieTable", () => {
  beforeEach(() => {
    mockGetData.mockImplementation((path: string) => {
      if (path.includes("movies")) {
        return mockMoviesData;
      }

      if (path.includes("movieCompanies")) {
        return mockMovieCompanyData;
      }
    });
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  test("Should match default snapshot", async () => {
    const { asFragment } = render(<MovieTable />);

    await act(wait);

    expect(asFragment()).toMatchSnapshot();
  });

  test("Should match loading state", async () => {
    const { getByTestId } = render(<MovieTable />);

    expect(getByTestId("loading-spinner")).toBeVisible();
  });

  test("Should match error state", async () => {
    const mockError = new Error("Something went wrong");

    mockGetData.mockImplementation(({ path }) => {
      if (path.includes("movies")) {
        return mockMoviesData;
      }

      if (path.includes("movieCompanies")) {
        throw mockError;
      }
    });

    const { getByTestId } = render(<MovieTable />);

    await act(wait);

    expect(getByTestId("error-paper")).toBeVisible();
  });

  test("CLICK - Should refetch movies when clicking refresh icon", async () => {
    const { getByTestId } = render(<MovieTable />);

    await act(wait);

    const refreshButton = getByTestId("refresh-button");

    act(() => fireEvent.click(refreshButton));

    await act(wait);

    expect(mockGetData).toHaveBeenCalledTimes(4);
  });

  test("CLICK - Should sort row when clicking score header", async () => {
    const { getByTestId } = render(<MovieTable />);

    await act(wait);

    const sortScore = getByTestId("sort-movie-table");

    // Ascending order
    act(() => fireEvent.click(sortScore));
    expect(getByTestId("movie-0-score").innerHTML).toEqual("0.2");

    // Descending Order
    act(() => fireEvent.click(sortScore));
    expect(getByTestId("movie-0-score").innerHTML).toEqual("1.3");
  });
});
