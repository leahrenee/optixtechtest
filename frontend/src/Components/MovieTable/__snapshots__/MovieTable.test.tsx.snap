// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MovieTable Should match default snapshot 1`] = `
<DocumentFragment>
  <div
    class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 css-192owgk-MuiPaper-root"
  >
    <div
      class="movie-table-header MuiBox-root css-0"
    >
      <p
        class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
      >
        Found 5 Movies
      </p>
      <button
        class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorPrimary MuiIconButton-sizeMedium css-1kuq5xv-MuiButtonBase-root-MuiIconButton-root"
        data-testid="refresh-button"
        tabindex="0"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
          data-testid="RefreshIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M17.65 6.35C16.2 4.9 14.21 4 12 4c-4.42 0-7.99 3.58-7.99 8s3.57 8 7.99 8c3.73 0 6.84-2.55 7.73-6h-2.08c-.82 2.33-3.04 4-5.65 4-3.31 0-6-2.69-6-6s2.69-6 6-6c1.66 0 3.14.69 4.22 1.78L13 11h7V4z"
          />
        </svg>
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </div>
    <div
      class="MuiTableContainer-root movie-table-container css-rorn0c-MuiTableContainer-root"
    >
      <table
        class="MuiTable-root css-rqglhn-MuiTable-root"
      >
        <thead
          class="MuiTableHead-root css-15wwp11-MuiTableHead-root"
        >
          <tr
            class="MuiTableRow-root MuiTableRow-head css-1q1u3t4-MuiTableRow-root"
          >
            <th
              class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-1ygcj2i-MuiTableCell-root"
              scope="col"
            >
              Title
            </th>
            <th
              class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-1ygcj2i-MuiTableCell-root"
              scope="col"
            >
              <span
                class="MuiButtonBase-root MuiTableSortLabel-root css-1qgma8u-MuiButtonBase-root-MuiTableSortLabel-root"
                data-testid="sort-movie-table"
                role="button"
                tabindex="0"
              >
                Review Score
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiTableSortLabel-icon MuiTableSortLabel-iconDirectionAsc css-1vweko9-MuiSvgIcon-root-MuiTableSortLabel-icon"
                  data-testid="ArrowDownwardIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M20 12l-1.41-1.41L13 16.17V4h-2v12.17l-5.58-5.59L4 12l8 8 8-8z"
                  />
                </svg>
              </span>
            </th>
            <th
              class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-1ygcj2i-MuiTableCell-root"
              scope="col"
            >
              Film Company
            </th>
          </tr>
        </thead>
        <tbody
          class="MuiTableBody-root css-apqrd9-MuiTableBody-root"
        >
          <tr
            class="MuiTableRow-root MuiTableRow-hover css-10cbwbc-MuiTableRow-root"
            data-testid="movie-0-row"
          >
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
            >
              A long train ride
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
              data-testid="movie-0-score"
            >
              1.3
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
            >
              Test Company 1
            </td>
          </tr>
          <tr
            class="MuiTableRow-root MuiTableRow-hover css-10cbwbc-MuiTableRow-root"
            data-testid="movie-1-row"
          >
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
            >
              Flowers on the meadow
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
              data-testid="movie-1-score"
            >
              0.6
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
            />
          </tr>
          <tr
            class="MuiTableRow-root MuiTableRow-hover css-10cbwbc-MuiTableRow-root"
            data-testid="movie-2-row"
          >
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
            >
              Summer
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
              data-testid="movie-2-score"
            >
              0.3
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
            >
              Test Company 1
            </td>
          </tr>
          <tr
            class="MuiTableRow-root MuiTableRow-hover css-10cbwbc-MuiTableRow-root"
            data-testid="movie-3-row"
          >
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
            >
              Back to the garden
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
              data-testid="movie-3-score"
            >
              0.6
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
            />
          </tr>
          <tr
            class="MuiTableRow-root MuiTableRow-hover css-10cbwbc-MuiTableRow-root"
            data-testid="movie-4-row"
          >
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
            >
              Mr John Smith
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
              data-testid="movie-4-score"
            >
              0.2
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
            />
          </tr>
        </tbody>
      </table>
    </div>
  </div>
</DocumentFragment>
`;
